File: src/ackward/logging/LoggerAdapter.cpp
#include <ackward/logging/LoggerAdapter.hpp>

#include <boost/python/dict.hpp>

#include <ackward/logging/Logger.hpp>
#include <ackward/core/PythonModule.hpp>

using namespace boost::python;

namespace ackward { namespace logging
{

LoggerAdapter::LoggerAdapter(const Logger& logger,
                             boost::python::dict extra) :
    Object (
        core::module().attr("LoggerAdapter")(
            logger, extra) )
{}

void LoggerAdapter::debug(const std::wstring& msg) const
{
    obj().attr("debug")(msg);
}

void LoggerAdapter::info(const std::wstring& msg) const
{
    obj().attr("info")(msg);
}

void LoggerAdapter::warning(const std::wstring& msg) const
{
    obj().attr("warning")(msg);
}

void LoggerAdapter::error(const std::wstring& msg) const
{
    obj().attr("error")(msg);
}

void LoggerAdapter::critical(const std::wstring& msg) const
{
    obj().attr("critical")(msg);
}

void LoggerAdapter::log(Level l, const std::wstring& msg) const
{
    obj().attr("log")(l, msg);
}

void LoggerAdapter::process(const std::wstring& msg,
                            dict kwargs) const
{
    obj().attr("process")(msg, kwargs);
}

}}

File: src/ackward/logging/LoggerAdapter.py
from ackward import (Class,
                     Constructor,
                     method,
                     Namespace,
                     TranslationUnit)

def tunit():
    return TranslationUnit(
        forward_declarations=[
            ('boost', 'python', 'class dict'),
            ('ackward', 'logging', 'class Logger'),
            ],
        header_includes=[
            ('string',),
            ('ackward', 'logging', 'Types.hpp'),
            ],
        impl_includes=[
            ('boost', 'python', 'dict.hpp'),
            ('ackward', 'logging', 'LoggerAdapter.hpp'),
            ('ackward', 'logging', 'Logger.hpp')])

def methods(parent):
    m = [
        'void debug(std::wstring msg) const',
        'void info(std::wstring msg) const',
        'void warning(std::wstring msg) const',
        'void error(std::wstring msg) const',
        'void critical(std::wstring msg) const',
        'void log(Level l, std::wstring msg) const',
        'void process(std::wstring msg, boost::python::dict kwargs) const',
        ]
    list(map(lambda m: method(m, parent=parent), m))

def definition(env):
    t = tunit()
    ns = Namespace('ackward', 'logging', parent=t)
    cls = Class(name='LoggerAdapter',
                wrapped_class='logging.LoggerAdapter',
                parent=ns)
    Constructor(
        signature=[('Logger', 'l'),
                   ('boost::python::dict', 'd')],
        parent=cls)
    methods(cls)
    return t
